diff --git a/riegeli/digests/BUILD b/riegeli/digests/BUILD
index cd741a0..c759927 100644
--- a/riegeli/digests/BUILD
+++ b/riegeli/digests/BUILD
@@ -83,8 +83,8 @@ cc_library(
     deps = [
         ":digester",
         "//riegeli/base:arithmetic",
+        "@com_google_absl//absl/crc:crc32c",
         "@com_google_absl//absl/strings",
-        "@crc32c",
     ],
 )
 
diff --git a/riegeli/digests/crc32c_digester.h b/riegeli/digests/crc32c_digester.h
index c0ff54f..4766bac 100644
--- a/riegeli/digests/crc32c_digester.h
+++ b/riegeli/digests/crc32c_digester.h
@@ -17,8 +17,8 @@
 
 #include <stdint.h>
 
+#include "absl/crc/crc32c.h"
 #include "absl/strings/string_view.h"
-#include "crc32c/crc32c.h"
 #include "riegeli/base/arithmetic.h"
 #include "riegeli/digests/digester.h"
 
@@ -72,8 +72,7 @@ constexpr uint32_t UnmaskCrc32c(uint32_t masked) {
 // Implementation details follow.
 
 inline void Crc32cDigester::WriteImpl(absl::string_view src) {
-  crc_ = crc32c::Extend(crc_, reinterpret_cast<const uint8_t*>(src.data()),
-                        src.size());
+  crc_ = static_cast<uint32_t>(absl::ExtendCrc32c(absl::crc32c_t{crc_}, src));
 }
 
 }  // namespace riegeli
