diff -Nur a/bazel/grpc_build_system.bzl b/bazel/grpc_build_system.bzl
--- a/bazel/grpc_build_system.bzl	2024-08-23 23:09:41.000000000 +0000
+++ b/bazel/grpc_build_system.bzl	2024-09-21 06:45:31.065224006 +0000
@@ -27,8 +27,8 @@
 Contains macros used throughout the repo.
 """
 
-load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test")
-load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
+# load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test")
+# load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
 load("@com_google_protobuf//bazel:upb_proto_library.bzl", "upb_proto_library", "upb_proto_reflection_library")
 load("//bazel:cc_grpc_library.bzl", "cc_grpc_library")
 load("//bazel:copts.bzl", "GRPC_DEFAULT_COPTS")
@@ -41,6 +41,14 @@
 # The set of known EventEngines to test
 EVENT_ENGINES = {"default": {"tags": []}}
 
+
+def ios_unit_test(**kwargs):
+    pass
+
+def ios_test_runner(**kwargs):
+    pass
+
+
 def if_not_windows(a):
     return select({
         "//:windows": [],
@@ -57,6 +65,12 @@
         "//conditions:default": [],
     })
 
+def if_mingw(a):
+    return select({
+        "//:mingw-gcc": a,
+        "//conditions:default": [],
+    })
+
 def _get_external_deps(external_deps):
     ret = []
     for dep in external_deps:
@@ -116,7 +130,7 @@
         "grpcpp_gcp_observability": PUBLIC,
         "grpc_resolver_fake": PRIVATE,
         "grpc++_public_hdrs": PUBLIC,
-        "grpc++_test": PRIVATE,
+        "grpc++_test": PUBLIC,
         "http": PRIVATE,
         "httpcli": PRIVATE,
         "iomgr_internal_errqueue": PRIVATE,
@@ -188,7 +202,7 @@
     copts = []
     if language.upper() == "C":
         copts = copts + if_not_windows(["-std=c11"])
-    linkopts = linkopts + if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"])
+    linkopts = linkopts + if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"]) + if_mingw(["-lws2_32"])
     if select_deps:
         for select_deps_entry in select_deps:
             deps += select(select_deps_entry)
@@ -217,8 +231,6 @@
         linkopts = linkopts,
         includes = [
             "include",
-            "src/core/ext/upb-gen",  # Once upb code-gen issue is resolved, remove this.
-            "src/core/ext/upbdefs-gen",  # Once upb code-gen issue is resolved, remove this.
         ],
         alwayslink = alwayslink,
         data = data,
@@ -549,7 +561,7 @@
     test_args = {
         "data": data,
         "copts": GRPC_DEFAULT_COPTS + copts,
-        "linkopts": if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"]),
+        "linkopts": if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"]) + if_mingw(["-lws2_32"]),
         "size": size,
         "timeout": timeout,
         "exec_compatible_with": exec_compatible_with,
diff -Nur a/BUILD b/BUILD
--- a/BUILD	2024-08-23 23:09:41.000000000 +0000
+++ b/BUILD	2024-09-21 00:03:07.146136113 +0000
@@ -25,6 +25,8 @@
     "grpc_upb_proto_library",
     "grpc_upb_proto_reflection_library",
     "python_config_settings",
+    "if_windows",
+    "if_mingw",
 )
 
 licenses(["reciprocal"])
@@ -205,6 +207,13 @@
 )
 
 config_setting(
+    name = "mingw-gcc",
+    flag_values = {
+        "@bazel_tools//tools/cpp:compiler": "mingw-gcc",
+    },
+)
+
+config_setting(
     name = "mac_x86_64",
     values = {"cpu": "darwin_x86_64"},
 )
@@ -2614,6 +2623,7 @@
     external_deps = [
         "protobuf_headers",
         "protobuf",
+        "absl/status",
     ],
     language = "c++",
     public_hdrs = [
@@ -3332,6 +3342,7 @@
         "absl/log:log",
         "absl/strings",
         "absl/strings:str_format",
+        "absl/types:optional",
     ],
     tags = ["nofixdeps"],
     deps = [
@@ -4117,6 +4128,7 @@
         "absl/log:log",
         "absl/strings",
     ],
+    linkopts = if_windows(["-defaultlib:crypt32.lib"]) + if_mingw(["-lcrypt32"]),
     language = "c++",
     visibility = ["@grpc:public"],
     deps = [
@@ -4447,6 +4459,7 @@
         "protobuf_struct_upb",
         "protobuf_timestamp_upb",
         "ref_counted_ptr",
+        "server_address",
         "uri_parser",
         "work_serializer",
         "//src/core:default_event_engine",
diff -Nur a/src/core/BUILD b/src/core/BUILD
--- a/src/core/BUILD	2024-08-23 23:09:41.000000000 +0000
+++ b/src/core/BUILD	2024-09-21 00:03:07.146136113 +0000
@@ -2632,6 +2632,7 @@
         "absl/log:log",
         "absl/strings",
         "absl/strings:str_format",
+        "absl/status",
     ],
     deps = [
         "event_engine_common",
