diff -Nur a/bazel/grpc_build_system.bzl b/bazel/grpc_build_system.bzl
--- a/bazel/grpc_build_system.bzl	2025-10-18 02:30:24.000000000 +0000
+++ b/bazel/grpc_build_system.bzl	2025-11-03 16:40:43.369920060 +0000
@@ -27,9 +27,9 @@
 Contains macros used throughout the repo.
 """
 
-load("@build_bazel_apple_support//rules:universal_binary.bzl", "universal_binary")
-load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test")
-load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
+# load("@build_bazel_apple_support//rules:universal_binary.bzl", "universal_binary")
+# load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test")
+# load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
 load("@com_google_protobuf//bazel:cc_proto_library.bzl", "cc_proto_library")
 load("@com_google_protobuf//bazel:upb_proto_library.bzl", "upb_proto_library", "upb_proto_reflection_library")
 load("@rules_proto//proto:defs.bzl", "proto_library")
@@ -45,6 +45,14 @@
 # The set of known EventEngines to test
 EVENT_ENGINES = {"default": {"tags": []}}
 
+
+def ios_unit_test(**kwargs):
+    pass
+
+def ios_test_runner(**kwargs):
+    pass
+
+
 def if_not_windows(a):
     return select({
         "//:windows": [],
@@ -59,6 +67,12 @@
         "//conditions:default": [],
     })
 
+def if_mingw(a):
+    return select({
+        "//:mingw-gcc": a,
+        "//conditions:default": [],
+    })
+
 def _get_external_deps(external_deps):
     ret = []
     for dep in external_deps:
@@ -144,7 +158,7 @@
     """
     visibility = _update_visibility(visibility)
     copts = []
-    linkopts = linkopts + if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"])
+    linkopts = linkopts + if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"]) + if_mingw(["-lws2_32"])
     if select_deps:
         for select_deps_entry in select_deps:
             deps += select(select_deps_entry)
@@ -180,8 +194,8 @@
         linkopts = linkopts,
         includes = [
             include_prefix + "include",
-            include_prefix + "src/core/ext/upb-gen",  # Once upb code-gen issue is resolved, remove this.
-            include_prefix + "src/core/ext/upbdefs-gen",  # Once upb code-gen issue is resolved, remove this.
+            # include_prefix + "src/core/ext/upb-gen",  # Once upb code-gen issue is resolved, remove this.
+            # include_prefix + "src/core/ext/upbdefs-gen",  # Once upb code-gen issue is resolved, remove this.
         ],
         alwayslink = alwayslink,
         data = data,
@@ -195,22 +209,7 @@
         srcs = srcs,
         deps = deps,
     )
-    universal_binary(
-        name = name + "_universal",
-        binary = name + "_native",
-    )
-
-    # In order to avoid warnings from Bazel, names of the rule and its output file must differ.
-    native.genrule(
-        name = name,
-        srcs = select({
-            "@platforms//os:macos": [name + "_universal"],
-            "//conditions:default": [name + "_native"],
-        }),
-        outs = [name + "_binary"],
-        cmd = "cp $< $@",
-        executable = True,
-    )
+    native.alias(name = name, actual = ":" + name + "_native")
 
 def grpc_internal_proto_library(
         name,
@@ -566,7 +565,7 @@
     test_args = {
         "data": data,
         "copts": GRPC_DEFAULT_COPTS + copts,
-        "linkopts": if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"]),
+        "linkopts": if_not_windows(["-pthread"]) + if_windows(["-defaultlib:ws2_32.lib"]) + if_mingw(["-lws2_32"]),
         "size": size,
         "timeout": timeout,
         "exec_compatible_with": exec_compatible_with,
diff -Nur a/BUILD b/BUILD
--- a/BUILD	2025-10-18 02:30:24.000000000 +0000
+++ b/BUILD	2025-11-03 16:32:32.746355607 +0000
@@ -26,6 +26,8 @@
     "grpc_upb_proto_library",
     "grpc_upb_proto_reflection_library",
     "python_config_settings",
+    "if_windows",
+    "if_mingw",
 )
 
 licenses(["reciprocal"])
@@ -260,6 +262,13 @@
     values = {"define": "use_strict_warning_windows=true"},
 )
 
+config_setting(
+    name = "mingw-gcc",
+    flag_values = {
+        "@bazel_tools//tools/cpp:compiler": "mingw-gcc",
+    },
+)
+
 python_config_settings()
 
 # This should be updated along with build_handwritten.yaml
@@ -3485,6 +3494,7 @@
         "absl/log:log",
         "absl/strings",
         "absl/strings:str_format",
+        "absl/types:optional",  # iomgr_timer
     ],
     tags = ["nofixdeps"],
     deps = [
@@ -4264,6 +4274,7 @@
         "absl/log:log",
         "absl/strings",
     ],
+    linkopts = if_windows(["-defaultlib:crypt32.lib"]) + if_mingw(["-lcrypt32"]),  # grpc_credentials_util
     visibility = ["//visibility:public"],
     deps = [
         "config_vars",
@@ -4589,6 +4600,7 @@
         "protobuf_struct_upb",
         "protobuf_timestamp_upb",
         "ref_counted_ptr",
+        "server_address",  # xds_client
         "uri",
         "work_serializer",
         "//src/core:down_cast",
diff -Nur a/src/core/BUILD b/src/core/BUILD
--- a/src/core/BUILD	2025-10-18 02:30:24.000000000 +0000
+++ b/src/core/BUILD	2025-11-03 16:32:32.750355604 +0000
@@ -2949,6 +2949,7 @@
     external_deps = [
         "absl/strings",
         "absl/strings:str_format",
+        "absl/status",
     ],
     deps = [
         "error",
