--- a/src/compiler/cpp_generator.cc	2025-04-21 22:04:38.000000000 +0000
+++ b/src/compiler/cpp_generator.cc	2025-05-02 07:47:05.531207288 +0000
@@ -1757,7 +1757,10 @@
         "grpcpp/impl/server_callback_handlers.h",
         "grpcpp/server_context.h",
         "grpcpp/impl/service_type.h",
-        "grpcpp/support/sync_stream.h"};
+        "grpcpp/support/sync_stream.h",
+         // ports_def.inc Must be included at last
+        "grpcpp/ports_def.inc",
+      };
     std::vector<std::string> headers(headers_strs, array_end(headers_strs));
     PrintIncludes(printer.get(), headers, params.use_system_headers,
                   params.grpc_search_path);
