diff -u a/bazel/upb_proto_library.bzl b/bazel/upb_proto_library.bzl
--- a/bazel/upb_proto_library.bzl	2023-05-23 23:37:19.000000000 +0000
+++ b/bazel/upb_proto_library.bzl	2023-05-24 01:49:45.036938849 +0000
@@ -249,6 +249,7 @@
             ]),
             progress_message = "Generating thunks for upb protos API for: " + ctx.label.name,
             mnemonic = "GenUpbProtosThunks",
+            use_default_shell_env = True,
         )
     return GeneratedSrcsInfo(
         srcs = srcs,
