diff -Nur b/bazel/java_lite_proto_library.bzl a/bazel/java_lite_proto_library.bzl
--- b/bazel/java_lite_proto_library.bzl	2025-10-15 18:33:29.000000000 +0000
+++ a/bazel/java_lite_proto_library.bzl	2025-11-03 16:28:32.702575230 +0000
@@ -9,5 +9,4 @@
 """java_lite_proto_library rule"""
 
-load("//bazel/private:java_lite_proto_library.bzl", _java_lite_proto_library = "java_lite_proto_library")
-
-java_lite_proto_library = _java_lite_proto_library
+def java_lite_proto_library(**kwattrs):
+    pass
diff -Nur b/bazel/java_proto_library.bzl a/bazel/java_proto_library.bzl
--- b/bazel/java_proto_library.bzl	2025-10-15 18:33:29.000000000 +0000
+++ a/bazel/java_proto_library.bzl	2025-11-03 16:28:25.458581961 +0000
@@ -9,5 +9,4 @@
 """java_proto_library rule."""
 
-load("//bazel/private:java_proto_library.bzl", _java_proto_library = "java_proto_library")
-
-java_proto_library = _java_proto_library
+def java_proto_library(**kwattrs):
+    pass
