diff --git a/BUILD.bazel b/BUILD.bazel
index a0c93676a..82d2a741b 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4,7 +4,6 @@
 load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_library", "cc_test")
 load("@rules_pkg//:mappings.bzl", "pkg_files", "strip_prefix")
 load("@rules_proto//proto:defs.bzl", "proto_lang_toolchain", "proto_library")
-load("@rules_python//python:defs.bzl", "py_library")
 load("@rules_java//java:defs.bzl", "java_binary", "java_lite_proto_library", "java_proto_library")
 load("//build_defs:cpp_opts.bzl", "COPTS", "LINK_OPTS", "PROTOC_LINK_OPTS")
 load(
 
diff --git a/protobuf.bzl b/protobuf.bzl
index 971612812..7692b0a77 100644
--- a/protobuf.bzl
+++ b/protobuf.bzl
@@ -1,7 +1,9 @@
 load("@bazel_skylib//lib:versions.bzl", "versions")
 load("@rules_cc//cc:defs.bzl", "cc_library")
 load("@rules_proto//proto:defs.bzl", "ProtoInfo")
-load("@rules_python//python:defs.bzl", "py_library", "py_test")
+
+py_library = native.py_library
+py_test = native.py_test

 def _GetPath(ctx, path):
     if ctx.label.workspace_root:
